clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp

```diff
@@ -0,0 +1,120 @@
+//===--- IncorrectEnableSharedFromThisCheck.cpp - clang-tidy --------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "IncorrectEnableSharedFromThisCheck.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/DeclCXX.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/Basic/Specifiers.h"
+#include "llvm/ADT/SmallPtrSet.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang::tidy::bugprone {
+
+void IncorrectEnableSharedFromThisCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(translationUnitDecl(), this);
+}
+
+void IncorrectEnableSharedFromThisCheck::check(
+    const MatchFinder::MatchResult &Result) {
+
+  class Visitor : public RecursiveASTVisitor<Visitor> {
+    IncorrectEnableSharedFromThisCheck &Check;
+    llvm::SmallPtrSet<const CXXRecordDecl *, 16> EnableSharedClassSet;
+
+  public:
+    explicit Visitor(IncorrectEnableSharedFromThisCheck &Check)
+        : Check(Check) {}
+
+    bool VisitCXXRecordDecl(CXXRecordDecl *RDecl) {
+      for (const auto &Base : RDecl->bases()) {
+        VisitCXXBaseSpecifier(Base, RDecl);
+      }
+      for (const auto &Base : RDecl->bases()) {
+        const CXXRecordDecl *BaseType = Base.getType()->getAsCXXRecordDecl();
+        if (BaseType && BaseType->getQualifiedNameAsString() ==
+                            "std::enable_shared_from_this") {
+          EnableSharedClassSet.insert(RDecl->getCanonicalDecl());
+          return true;
+        }
+      }
+      return true;
+    }
+
+    bool VisitCXXBaseSpecifier(const CXXBaseSpecifier &Base,
```

Comment:
this entire function is too complicated to what it does, for example you got multiple compares to std::enable_shared_from_this, or :enable_shared_from_this, that should be eliminated. 
And instead of using getQualifiedNameAsString, [isInStdNamespace](https://clang.llvm.org/doxygen/classclang_1_1Decl.html#a066b012f94431b5bba21d19715a274f4) and getName should be used.